### DISCUSSION TITLE
getting lot of undefined reference errors while installing latest version of curl.

### DISCUSSION THREAD

**Discussion by sh23k4r** [curl/curl#13981] [Q&A] [answered]
I downloaded the latest tar file from curl official site and trying it build it. I followed these steps :

cd curl-8.8.0 sudo ./configure --with-openssl sudo make

but after I do sudo make at last I get these errors :

`make[2]: Entering directory '/usr/src/curl-8.8.0/src'
CCLD     curl
../lib/.libs/libcurl.so: undefined reference to `EVP_PKEY_get_group_name'
../lib/.libs/libcurl.so: undefined reference to `SSL_get0_group_name'
../lib/.libs/libcurl.so: undefined reference to `EVP_PKEY_get_security_bits'
../lib/.libs/libcurl.so: undefined reference to `X509_STORE_load_file'
../lib/.libs/libcurl.so: undefined reference to `EVP_PKEY_get_id'
../lib/.libs/libcurl.so: undefined reference to `EVP_PKEY_get_bits'
../lib/.libs/libcurl.so: undefined reference to `SSL_get1_peer_certificate'
../lib/.libs/libcurl.so: undefined reference to `X509_STORE_load_path'
../lib/.libs/libcurl.so: undefined reference to `EVP_PKEY_get_bn_param'
../lib/.libs/libcurl.so: undefined reference to `EVP_PKEY_get0_type_name'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:1048: curl] Error 1
make[2]: Leaving directory '/usr/src/curl-8.8.0/src'
make[1]: *** [Makefile:1553: all-recursive] Error 1
make[1]: Leaving directory '/usr/src/curl-8.8.0/src'
make: *** [Makefile:1224: all-recursive] Error 1`
after a little searching on google, found out that it could be happening due to the older versions of openssl, so even I updated the openssl to the latest one:

`$ openssl --version
OpenSSL 3.3.1 4 Jun 2024 (Library: OpenSSL 3.3.1 4 Jun 2024)`
Host details :

`$ uname -a
Linux alpps789 4.18.0-513.24.1.el8_9.x86_64 #1 SMP Thu Mar 14 14:20:09 EDT 2024 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/redhat-release
Red Hat Enterprise Linux release 8.10 (Ootpa)`
Can anyone help me understand why this error is being thrown for libcurl.so and what steps to take to fix it?

I tried updating the openssl version as most of the communities saying its problem with older version of openssl.

**Comment by icing** [comment, marked as answer]
The openssl include files that cur configure finds are from OpenSSL 3.3.1, while the `libssl` that is linked is an older one. Therefore the compiler expects functions in `libssl` that simply are not there.

This looks like a confusing installation and would be a problem for any build on your machine to try to use openssl.

As free advice: building with `sudo` is not a good idea. if you do a `sudo make install` is will overwrite system packages and, should you ever abort such a job, you can easily end up in an incomplete install where header files and libs don't match.

**Reply by sh23k4r** [reply]
but how do we link a new libssl to the openSSL. The version of libcurl.so file I see is libcurl.so.4.8.0. Whatever unreferenced errors we are getting is all the functions used in openssl.c script. and when this script is used in sudo make, its not able to find those functions already defined somewhere I believe. I'm using sudo because the location I'm installing is not accessible otherwise.

**Reply by sh23k4r** [reply]
`$ sudo ldd libcurl.so
        linux-vdso.so.1 (0x00007ffdbe7df000)
        libssh2.so.1 => /lib64/libssh2.so.1 (0x00007faac550c000)
        libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007faac5278000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007faac4d8d000)
        libzstd.so.1 => /lib64/libzstd.so.1 (0x00007faac4ae9000)
        libz.so.1 => /lib64/libz.so.1 (0x00007faac48d1000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007faac46b1000)
        libc.so.6 => /lib64/libc.so.6 (0x00007faac42db000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007faac40d7000)
        /lib64/ld-linux-x86-64.so.2 (0x00007faac59fe000)` This is my lld on libcurl.so which shows its linked to the odler libssl

**Reply by sh23k4r** [reply]
I was able to install curl without those reference errors, but still if I check curl version, it shows openSSL not available.`$ curl --version
curl 8.8.0 (x86_64-pc-linux-gnu) libcurl/8.8.0 OpenSSL/not available zlib/1.2.11 zstd/1.4.4` and I think because of this, curl http requests are not working.

**Reply by sh23k4r** [reply]
Thanks. It worked now. I pointed it to the new libssl. Everything working as expected.

**Comment by icing** [comment]
Thanks.
